Native support for JCommander

While trying to get `spring-shell-samples` to work with `spring-native` there's an error with startup which seems to be around reflective dance happening in `JCommanderParameterResolver`.

We can remove this particular command from samples to give time to make things work while getting most parts of a sample to work with graal.

```
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'exampleApplicationRunnerConfiguration': Unsatisfied dependency expressed through field 'shell'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'shell': Invocation of init method failed; nested exception is org.springframework.shell.ParameterResolverMissingException: No parameter resolver found for parameter with index 0 (named 'args') of public java.lang.String org.springframework.shell.samples.jcommander.JCommanderCommands.jcommander(org.springframework.shell.samples.jcommander.Args) 
	at org.springframework.aot.beans.factory.InjectedFieldResolver.resolve(InjectedFieldResolver.java:50) ~[na:na]
	at org.springframework.aot.beans.factory.InjectedElementResolver.resolve(InjectedElementResolver.java:19) ~[na:na]
	at org.springframework.aot.beans.factory.InjectedElementResolver.invoke(InjectedElementResolver.java:37) ~[na:na]
	at org.springframework.shell.samples.ContextBootstrapInitializer.lambda$registerExampleApplicationRunnerConfiguration$1(ContextBootstrapInitializer.java:16) ~[na:na]
	at org.springframework.aot.beans.factory.ThrowableFunction.apply(ThrowableFunction.java:18) ~[na:na]
	at org.springframework.aot.beans.factory.BeanDefinitionRegistrar.lambda$instanceSupplier$0(BeanDefinitionRegistrar.java:97) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainFromSupplier(AbstractAutowireCapableBeanFactory.java:1249) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1191) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[na:na]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[na:na]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:953) ~[na:na]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918) ~[na:na]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583) ~[na:na]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:730) ~[spring-shell-samples:2.6.2]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:412) ~[spring-shell-samples:2.6.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:302) ~[spring-shell-samples:2.6.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1301) ~[spring-shell-samples:2.6.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1290) ~[spring-shell-samples:2.6.2]
	at org.springframework.shell.samples.SpringShellSample.main(SpringShellSample.java:38) ~[spring-shell-samples:3.0.0-SNAPSHOT]
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'shell': Invocation of init method failed; nested exception is org.springframework.shell.ParameterResolverMissingException: No parameter resolver found for parameter with index 0 (named 'args') of public java.lang.String org.springframework.shell.samples.jcommander.JCommanderCommands.jcommander(org.springframework.shell.samples.jcommander.Args) 
	at org.springframework.aot.context.annotation.InitDestroyBeanPostProcessor.invokeInitMethod(InitDestroyBeanPostProcessor.java:79) ~[na:na]
	at org.springframework.aot.context.annotation.InitDestroyBeanPostProcessor.postProcessBeforeInitialization(InitDestroyBeanPostProcessor.java:65) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:440) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1796) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[na:na]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[na:na]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[na:na]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276) ~[na:na]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1389) ~[na:na]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1309) ~[na:na]
	at org.springframework.aot.beans.factory.InjectedFieldResolver.resolve(InjectedFieldResolver.java:43) ~[na:na]
	... 22 common frames omitted
Caused by: org.springframework.shell.ParameterResolverMissingException: No parameter resolver found for parameter with index 0 (named 'args') of public java.lang.String org.springframework.shell.samples.jcommander.JCommanderCommands.jcommander(org.springframework.shell.samples.jcommander.Args) 
	at org.springframework.shell.Shell.lambda$validateParameterResolvers$5(Shell.java:335) ~[spring-shell-samples:3.0.0-SNAPSHOT]
	at java.util.Optional.orElseThrow(Optional.java:408) ~[na:na]
	at org.springframework.shell.Shell.lambda$validateParameterResolvers$6(Shell.java:335) ~[spring-shell-samples:3.0.0-SNAPSHOT]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[na:na]
	at java.util.stream.IntPipeline$1$1.accept(IntPipeline.java:180) ~[na:na]
	at java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:104) ~[na:na]
	at java.util.Spliterator$OfInt.forEachRemaining(Spliterator.java:699) ~[na:na]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[na:na]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[na:na]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[na:na]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[na:na]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497) ~[na:na]
	at org.springframework.shell.Shell.validateParameterResolvers(Shell.java:325) ~[spring-shell-samples:3.0.0-SNAPSHOT]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[na:na]
	at org.springframework.shell.Shell.gatherMethodTargets(Shell.java:102) ~[spring-shell-samples:3.0.0-SNAPSHOT]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.aot.context.annotation.InitDestroyBeanPostProcessor.invokeMethod(InitDestroyBeanPostProcessor.java:137) ~[na:na]
	at org.springframework.aot.context.annotation.InitDestroyBeanPostProcessor.invokeInitMethod(InitDestroyBeanPostProcessor.java:76) ~[na:na]
	... 35 common frames omitted

```

--- a/spring-shell-core/src/main/java/org/springframework/shell/Shell.java
+++ b/spring-shell-core/src/main/java/org/springframework/shell/Shell.java
@@ -32,6 +32,8 @@ import javax.validation.Validator;
 import javax.validation.ValidatorFactory;
 
 import org.jline.utils.Signals;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.core.MethodParameter;
@@ -55,6 +57,7 @@ import org.springframework.util.ReflectionUtils;
  */
 public class Shell {
 
+	private final static Logger log = LoggerFactory.getLogger(Shell.class);
 	private final ResultHandlerService resultHandlerService;
 
 	/**
@@ -287,14 +290,17 @@ public class Shell {
 	 * parameters that could not be resolved
 	 */
 	private Object[] resolveArgs(Method method, List<String> wordsForArgs) {
+		log.debug("Resolving args {} {}", method, wordsForArgs);
 		List<MethodParameter> parameters = Utils.createMethodParameters(method).collect(Collectors.toList());
 		Object[] args = new Object[parameters.size()];
 		Arrays.fill(args, UNRESOLVED);
 		for (ParameterResolver resolver : parameterResolvers) {
+			log.debug("Resolving args with {}", resolver);
 			for (int argIndex = 0; argIndex < args.length; argIndex++) {
 				MethodParameter parameter = parameters.get(argIndex);
 				if (args[argIndex] == UNRESOLVED && resolver.supports(parameter)) {
 					args[argIndex] = resolver.resolve(parameter, wordsForArgs).resolvedValue();
+					log.debug("Resolved {} {} {} {}", method, args[argIndex], resolver, parameter);
 				}
 			}
 		}

--- a/spring-shell-jcommander-adapter/src/main/java/org/springframework/shell/jcommander/JCommanderParameterResolver.java
+++ b/spring-shell-jcommander-adapter/src/main/java/org/springframework/shell/jcommander/JCommanderParameterResolver.java
@@ -1,19 +1,18 @@
 /*
-	* Copyright 2015 the original author or authors.
-	*
-	* Licensed under the Apache License, Version 2.0 (the "License");
-	* you may not use this file except in compliance with the License.
-	* You may obtain a copy of the License at
-	*
-	*      https://www.apache.org/licenses/LICENSE-2.0
-	*
-	* Unless required by applicable law or agreed to in writing, software
-	* distributed under the License is distributed on an "AS IS" BASIS,
-	* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-	* See the License for the specific language governing permissions and
-	* limitations under the License.
-	*/
-
+ * Copyright 2015-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package org.springframework.shell.jcommander;
 
 import java.lang.annotation.Annotation;
@@ -34,6 +33,8 @@ import com.beust.jcommander.JCommander;
 import com.beust.jcommander.Parameter;
 import com.beust.jcommander.ParameterException;
 import com.beust.jcommander.ParametersDelegate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import org.springframework.beans.BeanUtils;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -56,6 +57,7 @@ import static org.springframework.shell.Utils.unCamelify;
  */
 public class JCommanderParameterResolver implements ParameterResolver {
 
+	private final static Logger log = LoggerFactory.getLogger(JCommanderParameterResolver.class);
 	private static final Collection<Class<? extends Annotation>> JCOMMANDER_ANNOTATIONS = Arrays.asList(Parameter.class,
 			DynamicParameter.class, ParametersDelegate.class);
 
@@ -71,27 +73,25 @@ public class JCommanderParameterResolver implements ParameterResolver {
 		AtomicBoolean isSupported = new AtomicBoolean(false);
 		Class<?> parameterType = parameter.getParameterType();
 
-		if (isLegalReflectiveAccess(parameterType)) {
-			ReflectionUtils.doWithFields(parameterType, field -> {
-				if (isLegalReflectiveAccess(field.getType())) {
-					ReflectionUtils.makeAccessible(field);
-					boolean hasAnnotation = Arrays.stream(field.getAnnotations())
-							.map(Annotation::annotationType)
-							.anyMatch(JCOMMANDER_ANNOTATIONS::contains);
-					isSupported.compareAndSet(false, hasAnnotation);
-				}
-			});
+		log.debug("isLegalReflectiveAccess before");
+		ReflectionUtils.doWithFields(parameterType, field -> {
+			ReflectionUtils.makeAccessible(field);
+			boolean hasAnnotation = Arrays.stream(field.getAnnotations())
+					.map(Annotation::annotationType)
+					.anyMatch(JCOMMANDER_ANNOTATIONS::contains);
+			isSupported.compareAndSet(false, hasAnnotation);
+			log.debug("isLegalReflectiveAccess fields {}", hasAnnotation);
+		});
 
-			ReflectionUtils.doWithMethods(parameterType, method -> {
-				if (isLegalReflectiveAccess(method.getDeclaringClass())) {
-					ReflectionUtils.makeAccessible(method);
-					boolean hasAnnotation = Arrays.stream(method.getAnnotations())
-							.map(Annotation::annotationType)
-							.anyMatch(Parameter.class::equals);
-					isSupported.compareAndSet(false, hasAnnotation);
-				}
-			});
-		}
+		ReflectionUtils.doWithMethods(parameterType, method -> {
+			ReflectionUtils.makeAccessible(method);
+			boolean hasAnnotation = Arrays.stream(method.getAnnotations())
+					.map(Annotation::annotationType)
+					.anyMatch(Parameter.class::equals);
+			isSupported.compareAndSet(false, hasAnnotation);
+			log.debug("isLegalReflectiveAccess methods {}", hasAnnotation);
+		});
+		log.debug("isLegalReflectiveAccess supports {}", isSupported.get());
 		return isSupported.get();
 	}
 
@@ -135,11 +135,6 @@ public class JCommanderParameterResolver implements ParameterResolver {
 		return Stream.concat(
 				jCommander.getParameters().stream(),
 				jCommander.getMainParameterValue() != null ? Stream.of(jCommander.getMainParameterValue()) : Stream.empty());
-	}
-
-	// Java 9+ warn if you try to reflect on JDK types
-	private static boolean isLegalReflectiveAccess(Class<?> clzz) {
-		return (!clzz.getName().startsWith("java"));
 	}
 
 	@Override

--- a/spring-shell-jcommander-adapter/src/main/resources/META-INF/native-image/reflect-config.json
+++ b/spring-shell-jcommander-adapter/src/main/resources/META-INF/native-image/reflect-config.json
@@ -1,0 +1,18 @@
+[
+  {
+    "name":"com.beust.jcommander.JCommander"
+  },
+  {
+    "name":"com.beust.jcommander.converters.IntegerConverter",
+    "queryAllDeclaredConstructors":true,
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.String"] }]
+  },
+  {
+    "name":"com.beust.jcommander.validators.NoValidator",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"com.beust.jcommander.validators.NoValueValidator",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  }
+]

--- a/spring-shell-samples/src/main/java/org/springframework/shell/samples/jcommander/Args.java
+++ b/spring-shell-samples/src/main/java/org/springframework/shell/samples/jcommander/Args.java
@@ -1,0 +1,55 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.shell.samples.jcommander;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.validation.constraints.Min;
+
+import com.beust.jcommander.Parameter;
+
+/**
+ * An example straight from the JCommander documentation.
+ *
+ * @author Eric Bottard
+ * @author Cédric Beust
+ */
+public class Args {
+
+	@Parameter
+	private List<String> parameters = new ArrayList<>();
+
+	@Min(3)
+	@Parameter(names = { "-log", "-verbose" }, description = "Level of verbosity")
+	private Integer verbose = 1;
+
+	@Parameter(names = "-groups", description = "Comma-separated list of group names to be run")
+	private String groups;
+
+	@Parameter(names = "-debug", description = "Debug mode")
+	private boolean debug = false;
+
+	@Override
+	public String toString() {
+		return "Args{" +
+			"parameters=" + parameters +
+			", verbose=" + verbose +
+			", groups='" + groups + '\'' +
+			", debug=" + debug +
+			'}';
+	}
+}

--- a/spring-shell-samples/src/main/java/org/springframework/shell/samples/jcommander/JCommanderCommands.java
+++ b/spring-shell-samples/src/main/java/org/springframework/shell/samples/jcommander/JCommanderCommands.java
@@ -1,0 +1,36 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.shell.samples.jcommander;
+
+import org.springframework.shell.standard.ShellComponent;
+import org.springframework.shell.standard.ShellMethod;
+import org.springframework.shell.standard.ShellOption;
+
+import javax.validation.Valid;
+
+/**
+ * A class with JCommander commands.
+ *
+ * @author Eric Bottard
+ */
+@ShellComponent
+public class JCommanderCommands {
+
+	@ShellMethod("Bind parameters to JCommander POJO.")
+	public String jcommander(@ShellOption(optOut = true) @Valid Args args) {
+		return "You said " + args;
+	}
+}

--- a/spring-shell-samples/src/main/resources/META-INF/native-image/reflect-config.json
+++ b/spring-shell-samples/src/main/resources/META-INF/native-image/reflect-config.json
@@ -1,0 +1,9 @@
+[
+  {
+    "name":"org.springframework.shell.samples.jcommander.Args",
+    "allDeclaredFields":true,
+    "queryAllDeclaredMethods":true,
+    "queryAllDeclaredConstructors":true,
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  }
+]

